storage: block preemptive snapshot on member replica

tbg opened the pull request:
This kind of thing would usually happen when we add a replica on S,
remove it, and re-add it quickly (before the replica GC queue has
removed the replica).

First of all, it sounds scary to fast-forward Raft state with
out-of-band snapshots, though it seems to work in practice.

Secondly, this behaves weirdly because it leaves the resulting Replica
at its previous replicaID. In particular, the preemptive snapshot is
completely unprotected from the replica GC queue, which is going to GC
the replica soon.

I ran into this running a small restore and noticing a Raft snapshot
that highlighted the above (in addition, the preemptive snapshot was
discarded by Raft because the replica had been caught up all the way to
the snapshot already, but I think that's irrelevant to what happened).

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/32436)
<!-- Reviewable:end -->

tbg commented:
Fixed this PR up, PTAL.

bdarnell reviewed (approved):
> Reviewed 3 of 3 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32436)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

benesch reviewed (approved):
> Reviewed 3 of 3 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32436)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/storage/store_snapshot.go, line 458 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/32436#-LSGDEDbDw0gjhJIHlWW:-LSGDEDbDw0gjhJIHlWX:b-dyi9ar) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb72459cddeb3c099e9da503519092feb93e676f/pkg/storage/store_snapshot.go#L458)):*
> ```Go
> 			//
> 			// In the happy case, the existing replica is uninitialized and the
> 			// overlap check below will return success (with a placeholder).
> ```

But we know it is initialized because we're in the existingIsInitialized branch? Seems like a strange place to describe the happy case.


<!-- Sent from Reviewable.io -->

tbg committed:
storage: block preemptive snapshot on member replica

Prior to this commit, a preemptive snapshot would be allowed to apply to
a replica that was (or at least believed itself to be) part of a Raft
group (i.e. nonzero ReplicaID).

This kind of thing would usually happen when we add a replica on S,
remove it, and re-add it quickly (before the replica GC queue has
removed the replica).

This seemed to work in practice (even though it seems weird and bypasses
the Raft leader's internal status tracking, which has theoretical
potential to trigger assertions), but had one downside: It leaves the
resulting Replica at its existing replicaID, making it more likely in
the up-down-up scenario above that the preemptive snapshot would be
removed along with the old incarnation of the replica, requiring yet
another Raft snapshot to restore the data.

I ran into this running a small restore and noticing a Raft snapshot
that highlighted the above (in addition, the preemptive snapshot was
discarded by Raft because the replica had been caught up all the way to
the snapshot already, but I think that's irrelevant to what happened).

Release note: None

tbg reviewed (commented):
Some more archaeology and minor changes, PTAL.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32436#-:-LSZttio74OrcrgF1KnG:bu0mhue)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

benesch reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 3 of 3 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32436#-:-LS_GLOY-iHSpFoYafPd:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

tbg commented:
bors r=benesch
🙇
